lighthouse-core/gather/gather-runner.js

```diff
@@ -337,52 +336,64 @@ class GatherRunner {
+   * @param {LH.BaseArtifacts} baseArtifacts
    * @return {Promise<LH.Artifacts>}
    */
-  static async collectArtifacts(gathererResults, tracingData, settings) {
-    // Can't handle dynamic GathererResults -> Artifacts, so explicitly make *
-    // and cast to Artifacts before returning.
-    /** @type {Object<string, *>} */
-    const artifacts = {
-      traces: tracingData.traces,
-      devtoolsLogs: tracingData.devtoolsLogs,
-      settings,
-      // Take only unique LighthouseRunWarnings, if any.
-      LighthouseRunWarnings: Array.from(new Set(gathererResults.LighthouseRunWarnings)),
-    };
+  static async collectArtifacts(gathererResults, baseArtifacts) {
+    /** @type {Partial<LH.GathererArtifacts>} */
+    const gathererArtifacts = {};
 
     const pageLoadFailures = [];
-    for (const [gathererName, phaseResultsPromises] of Object.entries(gathererResults)) {
-      if (artifacts[gathererName] !== undefined) continue;
+    const resultsEntries = /** @type {GathererResultsEntries} */ (Object.entries(gathererResults));
+    for (const [gathererName, phaseResultsPromises] of resultsEntries) {
+      if (gathererArtifacts[gathererName] !== undefined) continue;
 
       try {
         const phaseResults = await Promise.all(phaseResultsPromises);
         // Take last defined pass result as artifact.
         const definedResults = phaseResults.filter(element => element !== undefined);
         const artifact = definedResults[definedResults.length - 1];
-        artifacts[gathererName] = artifact;
+        // Typecast pretends artifact always provided here, but this is actually checked below.
+        gathererArtifacts[gathererName] = /** @type {NonVoid<PhaseResult>} */ (artifact);
       } catch (err) {
         // An error result must be non-fatal to not have caused an exit by now,
         // so return it to runner to handle turning it into an error audit.
-        artifacts[gathererName] = err;
+        gathererArtifacts[gathererName] = err;
         // Track page load errors separately, so we can fail loudly if needed.
         if (LHError.isPageLoadError(err)) pageLoadFailures.push(err);
       }
 
-      if (artifacts[gathererName] === undefined) {
+      if (gathererArtifacts[gathererName] === undefined) {
         throw new Error(`${gathererName} failed to provide an artifact.`);
       }
 
       // Fail the run if more than 50% of all artifacts failed due to page load failure.
-      if (pageLoadFailures.length > Object.keys(artifacts).length * 0.5) {
+      if (pageLoadFailures.length > Object.keys(gathererArtifacts).length * 0.5) {
         throw pageLoadFailures[0];
       }
     }
 
-    return /** @type {LH.Artifacts } */ (artifacts);
+    // Take only unique LighthouseRunWarnings.
+    baseArtifacts.LighthouseRunWarnings = Array.from(new Set(baseArtifacts.LighthouseRunWarnings));
+
+    return /** @type {LH.Artifacts} */ ({...baseArtifacts, ...gathererArtifacts});
```

Comment:
can drop this cast when computed artifacts aren't part of the `Artifacts` interface anymore